src/storage/mod.rs

```diff
@@ -0,0 +1,566 @@
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+    S::Key: Clone,
+{
+    let mut init = Some(init);
+    let signal = {
+        if cfg!(feature = "ssr") {
+            // SSR does not support synced storage on the backend. We will just use a normal Signal to represent the initial state.
+            // The client will hydrate this with a correct SyncedStorageEntry and maintain state.
+            Signal::new(init.take().unwrap()())
+        } else if cfg!(feature = "hydrate") {
+            let key_clone = key.clone();
+            let storage_entry = new_synced_storage_entry::<S, T>(key, init.take().unwrap());
+            if cx.generation() == 0 {
+                // The first generation is rendered on the server side and so must be hydrated.
+                cx.needs_update();
+            }
+            if cx.generation() == 1 {
+                // The first time the vdom is hydrated, we set the correct value from storage and set up the subscription to storage events.
+                storage_entry
+                    .entry
+                    .set(get_from_storage::<S, T>(key_clone, init.take().unwrap()));
+                storage_entry.save_to_storage_on_change();
+                storage_entry.subscribe_to_storage(cx);
+            }
+            *storage_entry.data()
+        } else {
+            // The client is rendered normally, so we can just use the synced storage entry.
+            let storage_entry = new_synced_storage_entry::<S, T>(key, init.take().unwrap());
+            storage_entry.save_to_storage_on_change();
+            storage_entry.subscribe_to_storage(cx);
+            *storage_entry.data()
+        }
+    };
+    signal
+}
+
+/// A hook that creates a StorageEntry with the latest value from storage or the init value if it doesn't exist.
+pub fn use_storage_entry<S, T>(
+    cx: &ScopeState,
+    key: S::Key,
+    init: impl FnOnce() -> T,
+) -> &mut StorageEntry<S, T>
+where
+    S: StorageBacking,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+    S::Key: Clone,
+{
+    cx.use_hook(|| new_storage_entry::<S, T>(key, init))
+}
+
+/// A hook that creates a StorageEntry with the latest value from storage or the init value if it doesn't exist, and provides a channel to subscribe to updates to the underlying storage.
+pub fn use_synced_storage_entry<S, T>(
+    cx: &ScopeState,
+    key: S::Key,
+    init: impl FnOnce() -> T,
+) -> &mut SyncedStorageEntry<S, T>
+where
+    S: StorageBacking + StorageSubscriber<S>,
+    T: Serialize + DeserializeOwned + Clone + Send + Sync + PartialEq + 'static,
+    S::Key: Clone,
+{
+    cx.use_hook(|| new_synced_storage_entry::<S, T>(key, init))
+}
+
+/// Returns a StorageEntry with the latest value from storage or the init value if it doesn't exist.
+pub fn new_storage_entry<S, T>(key: S::Key, init: impl FnOnce() -> T) -> StorageEntry<S, T>
```

Comment:
wondering if these and a lot of the following functions need to be public. Do we expect to allow other folks to directly extend dioxus-storage in their own repos or do we want them to contribute them back here?

Reply:
I'm using `new_storage` (formerly `storage`) inside of `use_context_provider` callback:

https://github.com/juspay/nix-browser/blob/bc09af706cc3ac9776080d25c5c1d9d38fb5e599/src/app/state.rs#L78

Reply:
I think we should keep them public. A crate that defines something related to storage may want to create their own storage hook. For example a crate that defines a serialization format or database wrapper may want to define a storage hook.